### PR TITLE
Add dense_matmul implementation for gpt-oss

### PR DESCRIPTION
**Pull request by RissyRan**
# Description

Appreciate great work from @shuningjin for verification!

* Add dense_matmul implementation
* Remove the assertion from pyconfig.py
* Add compile tests for functionality

# Tests

Test locally for logits verification:

```
gpt-oss-20b, max_target_length=max_prefill_predict_length=200

1 Dense + CPU
Max Numerical Difference 2.8133392333984375e-05
max KL divergence = 1.0624022195315774e-07

2 Dense + TPU
Max Numerical Difference 0.0010275840759277344
max KL divergence = 1.9383925575766625e-07
```

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: src/MaxText/layers/moe.py [hunk 1]
```diff
@@ -1635,13 +1639,13 @@ def dense_matmul(
...
-          layer_w1 = layer_w1 + w1_bias
+          layer_w1 = layer_w1 + w1_bias[None, None, :, :]
         if self.config.activations_in_float32:
           layer_w1 = layer_w1.astype(jnp.float32)
         layer_w1 = adc.checkpoint_name(layer_w1, "mlpwi_1")
       # pylint: disable=protected-access
-      layer_w0_act = linears._convert_to_activation_function(self.config.mlp_activations[0])(layer_w0)
-      layer_multiply = jnp.multiply(layer_w0_act, layer_w1).astype(self.dtype)
+      layer_multiply = self.apply_ffn_activation(layer_w0, layer_w1)
+
```
Comment (gagika) [inline src/MaxText/layers/moe.py:1650]: there is also similar code at line 1581 that you can call the new function `self.apply_ffn_activation`.

File: src/MaxText/layers/moe.py [hunk 2]
```diff
@@ -489,6 +489,19 @@ def deepseek_routing(self, gate_logits: jax.Array, pre_bias_logits: jax.Array) -
     top_k_weights = jnp.take_along_axis(pre_bias_logits, top_k_indices, axis=-1)
     return top_k_weights, top_k_indices
 
+  def apply_ffn_activation(self, layer_w0, layer_w1):
+    """Applies FFN activation function."""
+    if self.config.decoder_block == ctypes.DecoderBlockType.GPT_OSS:
+      layer_w0 = jnp.clip(layer_w0, a_min=None, a_max=self.config.mlp_activations_limit)
+      layer_w1 = jnp.clip(layer_w1, a_min=-self.config.mlp_activations_limit, a_max=self.config.mlp_activations_limit)
+      layer_act = linears._convert_to_activation_function(self.config.mlp_activations[0])(layer_w0 * 1.702)
```
Comment (gagika) [inline src/MaxText/layers/moe.py]: seams it's both in if and else, perhaps you can pull it out.

File: src/MaxText/layers/moe.py [hunk 3]
```diff
@@ -489,6 +489,19 @@ def deepseek_routing(self, gate_logits: jax.Array, pre_bias_logits: jax.Array) -
     top_k_weights = jnp.take_along_axis(pre_bias_logits, top_k_indices, axis=-1)
     return top_k_weights, top_k_indices
 
+  def apply_ffn_activation(self, layer_w0, layer_w1):
+    """Applies FFN activation function."""
+    if self.config.decoder_block == ctypes.DecoderBlockType.GPT_OSS:
```
Comment (gagika) [inline src/MaxText/layers/moe.py]: could you have something like this for profiling?

with jax.named_scope("ffn_act"):

Comment (RissyRan) [reply src/MaxText/layers/moe.py]: Thanks! Added.

File: src/MaxText/layers/moe.py (hunk 2 above)
Comment (RissyRan) [reply src/MaxText/layers/moe.py]: Updated.

File: src/MaxText/layers/moe.py (hunk 1 above)
Comment (RissyRan) [reply src/MaxText/layers/moe.py:1650]: Good catch!
